lib/CodeGen/CGExprScalar.cpp

```diff
@@ -1503,6 +1503,27 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
   case CK_LValueToRValue:
     assert(CGF.getContext().hasSameUnqualifiedType(E->getType(), DestTy));
     assert(E->isGLValue() && "lvalue-to-rvalue applied to r-value!");
+
+    if (CE->hasInferredBoundsExpr()) {
+      BoundsExpr* InferredBounds = CE->getInferredBoundsExpr();
+
+      if (InferredBounds->isInvalid())
+        break;
+
+      if (isa<ArraySubscriptExpr>(E)) {
+        CGF.SetNextBoundsCheckExpr(InferredBounds, CodeGenFunction::BC_ArraySubscript);
+      }
+      else if (const auto *UO = dyn_cast<UnaryOperator>(E)) {
+        switch (UO->getOpcode())
+        {
+        case UO_Deref:
+          CGF.SetNextBoundsCheckExpr(InferredBounds, CodeGenFunction::BC_Deref);
+        default:
```

Comment:
For the moment I am unwilling to put these asserts in, because I am currently explicitly only handling two kinds of inferred bounds. When I am using all the inferred bounds, or expect to be doing so, sure, for the moment it errors when I need it not to.